### PR TITLE
increase compatibility between BatMap, BatSplay and stdlib.Map

### PR DESCRIPTION
**Pull request by jabot2**
There were several functions added to stdlib.Map, which are added in this PR.
Unfortunately, stdlib also added an update function with an incompatible type signature, 
which is why this PR adds an update_stdlib function.

For 100% compatibility, something like `  module IntMap = struct
     include BatMap.Int
     let update = update_stdlib
  end
  ` is necessary, see also the extended batteries_compattest.

Also, for several functions the exception raised on empty maps was changed to `Not_found` to increase consistency and compatibility with stdlib; the exception raised was unspecified by the documentation for these functions anyways.

### CODE REVIEW

File: src/batSplay.ml
Comment (gasche) [inline src/batSplay.ml:810]: This is not a very nice implementation. The explanation is not correct (a splay tree is a mutable pointer to a persistent tree structure, so you could just dereference the tree and then iterate on it on-demand), but more importantly the work of providing an iterator on splay tree elements has already been done for the `Enum` datatype and it is easy to reuse it. Look at the type `'a enumeration`, the function `cons_enum` and the way it is used in the functions `equal` or `compare` (they traverse two trees, you only need one, but the idea is the same).

File: src/batSplay.ml
Comment (gasche) [inline src/batSplay.ml:814]: Again, this implementation is not satisfactory, it does not provide the expected complexity. Could you try implementing this by using a function that takes a bound `k` and a search tree `tr`, and returns an `'a enumeration` of the elements whose keys are greater or equal to `k`?

File: testsuite/test_map.ml
Comment (gasche) [inline testsuite/test_map.ml:83]: this is a nice trick to test this aspect of the Map module, thanks

**Review by gasche — COMMENTED** [review]
Thanks for the code! I made a first pass of review.

I regret not insisting further to use our proper enumeration type on sets before, because the same issue is showing up again. If you are sure that you don't have the time or knowledge to provide better implementations of these functions, feel free to say so, and I can have a look, but it's actually not so hard to use them, so I think you could give it a try.

File: src/batMap.mli
Comment (jabot2) [reply src/batMap.mli]: fixed

File: src/batMap.mli
Comment (jabot2) [reply src/batMap.mli]: done

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml]: Probably because my editor indents like this. Reverted...

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml:517]: Consistency with max_binding, choose, find (_first, _last), Consistency with BatMap and compatibility with the stdlib: https://ocaml.org/releases/4.11/htmlman/libref/Map.S.html

Note that the batteries documentation does not (yet) specify which exception is raised (which, again, should be fixed by this PR), see http://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatMap.html

Comment (jabot2) [reply src/batSplay.ml:517]: I mean, the current situation for batMap is:
- min_binding and any raise Not_found
- pop_min_binding and pop_max_binding have assert(false)
- max_binding has invalid_arg "PMap.max_binding: empty tree"
  (the error message seems... outdated?)
None of these exceptions are documented.

For BatSplay the current situation is:
- min_binding, any and max_binding raise Not_found
- pop_min_binding and pop_max_binding have assert(false)

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml:810]: I did not realize that it is possible to use the persistent tree structure. In hindsight it seems obvious...
I will rewrite this, i will probably copy the implementation from BatSet.to_seq.

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml:814]: Yes, i will re-use the implementation from BatMap.

File: src/batSplay.ml
Comment (gasche) [reply src/batSplay.ml]: > i will probably copy the implementation from BatSet.to_seq or BatMap.to_seq.

We designed a datatype (`'a enumeration`) specifically to provide an efficient solution to this problem, and I am not sure why you don't want to use it.

File: src/batSplay.ml
Comment (gasche) [reply src/batSplay.ml:517]: The authors of this code used `assert false` because they believe that this point can never be reached, for any input. I haven't looked at the code deeply, but I think that the reasoning is that `choose tr` was called first, so we already know that there is at least one element. If you find this unclear, I think it would be a good idea to add a comment. If this is wrong for some reason I missed, then indeed it needs to be changed. But if it is true that this case can never be reached, then I think `assert false` is the right choice.

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml]: So, i tried to implement to_seq with 'a enumeration and without:
```
  let to_seq m () = (* version with cons_enum *)
    let rec to_seq_hlp m =
      match m with
      | End -> BatSeq.Nil
      | More(k, v, r, e) ->
         BatSeq.Cons ((k, v), fun () -> to_seq_hlp (cons_enum r e)) in
    to_seq_hlp (cons_enum (sget m) End)

  let to_seq m = (*  version without cons_enum *)
    let rec to_seq_hlp m =
      fun () ->
      match m with
      | Empty -> BatSeq.Nil
      | Node(l, kv, r) ->
         BatSeq.append (to_seq_hlp l) (fun () -> BatSeq.Cons (kv, to_seq_hlp r)) () in
    to_seq_hlp (sget m)
```

I can see that `cons_enum` and the `'a enumeration` data type might be preferable to generating closures.
However, creating a `'a seq` requires creating one closure per item anyways, so we're not gaining anything.

For `to_seq_from` it gets worse: either we reimplement the `cons_enum` function to compare with the given key and skip descending into the left subtree, or we generat a full enumeration and throw away the leading N elements that are too small.

Comment (gasche) [reply src/batSplay.ml]: Let me first apologize for the somewhat grumpy tone of my last remark. I will look at using `'a enumeration` in BatSeq at first and see if we have a measurable performance difference, before we decide what to do for BatMap and BatSplay.

Comment (jabot2) [reply src/batSplay.ml]: I will try to reimplement using enumeration.

Comment (gasche) [reply src/batSplay.ml]: I proposed a benchmark for `BatSet.to_seq` in #1010. It seems that the enumeration/iteration-based implementation is sensibly faster. (I don't find it surprising, but it is better to have hard data than guesses.) I proposed an implementation, still for `BatSet`, in #1011; for `to_seq_from` I indeed used the approach you suggest of a new `cons_iter_from` function. Feel free to comment on the implementations, and/or to reuse the idea for Map/Splay if you find it convincing.

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml:517]: You're right, i missed that get_root (for batMap) and choose (for batSplay) are called before, and raise Not_found - so the assert(false) here should be unreachable.

That means that the only place where the "wrong" (in my opinion) exception is raised is in max_binding, which does not raise `Not_found`.

I will reinstate the `assert(false)` and add a comment.

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml]: After implementing to_seq and to_seq_from for BatSplay i also used the implementation for BatMap and BatSet.
You can see my implementations in the PR, they look remarkably similar to your implementation in the benchmark.

File: src/batSplay.ml
Comment (jabot2) [reply src/batSplay.ml]: ... is now implemented in terms of 'a enumeration

**Comment by jabot2** [comment]
Out of curiosity: what is the oldest ocaml version we still need to support with batteries-included?
Can we (e.g) presume that ppx rewriters are available?

File: src/batSet.ml
Comment (gasche) [inline src/batSet.ml:726]: Now we have two comparable implementations of `to_seq` and `to_seq_from`, this one and the one from #1011. Which one do we want to merge and reuse in the other modules?

My preference would go for merging @jabot2's code (it's always nicer to get code from fresh new contributors than grumpy old ones), but there are aspects of the implementation in #1011 that are better:
- I prefer defining `cons_enum_from` as a toplevel function, close to `cons_iter` (... which suggests that it should be called `cons_iter_from`), because it could be reused for other iterators later.
- currently `to_seq_hlp` returns a `'a BatSeq.node` and the caller takes an additional `()` argument, it would be more efficient if it returned a `'a BatSeq.t` directly. (The problem here is that OCaml assumes that `to_seq_from` is a 4-argument function, which get optimized differently from 3-argument functions.)

(I have a small preference for my naming choice, `seq_of_iter` rather than `to_seq_hlp`, but this is not very important and we could go with your choice.)

**Comment by gasche** [comment]
> Out of curiosity: what is the oldest ocaml version we still need to support with batteries-included?
> Can we (e.g) presume that ppx rewriters are available?

In general I think that the older version that it makes sense supporting is the OCaml version of the current Debian-stable release (or possibly oldstable if it's not too hard). Currently stable has 4.05 and oldstable has 4.02.3. Either of these would be fine.

(On the other hand, using ppx rewriters can introduce a fair bit of maintenance churn/pain, so this is something to decide very carefully.)

File: src/batSplay.ml
Comment (gasche) [reply src/batSplay.ml]: Thanks! (I agree that the exception for `max_binding` was wrong.)

File: src/batMap.ml
Comment (gasche) [inline src/batMap.ml:57]: This should be just `= Empty`. In general we don't want to use physical equality `==` unless when absolutely necessary, because it has a tricky semantics that is less portable and needs more careful code reading.

File: src/batMap.ml
Comment (gasche) [inline src/batMap.ml:580]: Why did you change the implementation here? If I understand correctly, the previous implementation is better when few keys remain in the result, and the new implementation is better when most keys remain in the result.

Remarks:
- it is odd that `filterv` and `filter` would use opposite algorithms
- we could implement a direct version using `bal` that should be fast in all cases (but then we have already spent a lot of time on this PR so maybe it's okay to keep less-good code if it was already there).

File: src/batMap.ml
Comment (gasche) [inline src/batMap.ml:925]: This implementation does not have the expected complexity. I think that you could just use `merge` instead, which is more general. (A hand-written implementation of `union` using the same approach as `merge` would also be faster.)

File: src/batMap.ml
Comment (gasche) [inline src/batMap.ml:1744]: This implementation is wrong if `m1.cmp` is not compatible with `m2.map`. See how `heuristic_merge` is implemented.

**Review by gasche — COMMENTED** [review]
I did a complete review pass over `batMap.ml`.

File: src/batMap.ml
Comment (jabot2) [reply src/batMap.ml:580]: Because if the callback `f` returns true for all bindings then the given map should be returned unmodified.
That behaviour cannot be achieved by starting from `empty` and adding the missing elements...

I can make the implementation of filterv and filter consistent though.

File: src/batMap.ml
Comment (jabot2) [reply src/batMap.ml:57]: ok

File: src/batMap.ml
Comment (jabot2) [reply src/batMap.ml:925]: i used heuristic_merge instead. That also avoids problems when PMaps with incompatible comparisons are merged (see your comment below)

File: src/batMap.ml
Comment (jabot2) [reply src/batMap.ml:1752]: avoided by implementing union_stdlib in terms of heuristic_merge instead

File: src/batSet.ml
Comment (gasche) [reply src/batSet.ml:727]: (The implementation was rewritten, taking all these points into account, so it is now perfectly fine.)

File: src/batSplay.ml
Comment (gasche) [inline src/batSplay.ml:837]: This implementation is not completely satisfying (it does not preserve the knowledge that we build the result in-order), but here the advice to use `merge` instead is not so convincing, because `merge` is okay but not that nice. (It performs an imperative traversal of two `Enum`, instead of working on the nicer, underlying `'a enumeration` structures.) I propose to leave it as-is in the PR, and I maybe submit a subsequent PR to improve `merge` and then define `union` from it.

**Review by gasche — APPROVED** [review]
I'm done reviewing the PR and I only had somewhat minor comments left. I believe that this is ready for merging; well done!

What kind of a git-history person are you? Would you like to rebase your commits before merging and merge some of them together, or leave things as-is, or that we squash all of them in a single commit?

File: src/batMap.ml
Comment (gasche) [reply src/batMap.ml:580]: I may rewrite `filter_map` to use ordered join operations instead of this unpleasant `fold`, and define `filter` and `filterv` on top of it. This can wait until the PR is merged.

**Comment by jabot2** [comment]
Considering git: If you prefer a "clean" history, feel free to squash the commits; I personally don't hold a strong opinion on the matter.

**Pull request merged**
